internal/driver/driver.go

```diff
@@ -0,0 +1,790 @@
+
+		var nodename = cast.ToString(req.Attributes["NodeName"])
+		var dbInfo, _ = s.getDBInfo(nodename)
+
+		reading, err := newCommandValue(req.Type, params[i])
+		if err != nil {
+			s.lc.Errorf("newCommandValue error: %s", err)
+		}
+		helper.SetValueAt(dataset[i], 0, reading)
+
+		// create gos7 DataItem
+		var s7DataItem = gos7.S7DataItem{
+			Area:     dbInfo.Area,
+			WordLen:  dbInfo.WordLength,
+			DBNumber: dbInfo.DBNumber,
+			Start:    dbInfo.Start,
+			Amount:   dbInfo.Amount,
+			Data:     dataset[i],
+			Error:    s7_errors[i],
+		}
+		s7DataItems = append(s7DataItems, s7DataItem)
+
+	}
+
+	s.lc.Debugf("Write to S7DataItems: %s", s7DataItems)
+
+	// send command requests
+	times := int(reqs_len / batch_size)
+	remains := reqs_len % batch_size
+	var tmp_reqs []sdkModel.CommandRequest
+	var tmp_s7DateItems = []gos7.S7DataItem{}
+
+	s7Client := s.getS7Client(deviceName, protocols)
+
+outloop:
+	for j := 0; j <= times; j++ {
+
+		tmp_s7DateItems = tmp_s7DateItems[:0] // clear array
+		tmp_reqs = tmp_reqs[:0]               // clear array
+		if j >= times {
+			if remains > 0 {
+				tmp_s7DateItems = s7DataItems[times*batch_size : times*batch_size+remains]
+			} else {
+				break outloop // load complete, exit loop
+			}
+		} else {
+			tmp_s7DateItems = s7DataItems[j*batch_size : j*batch_size+batch_size]
+		}
+
+		// write data to S7 device, if error, try 3 times
+		retrytimes := 3
+		for {
+			err = s7Client.Client.AGWriteMulti(tmp_s7DateItems, len(tmp_s7DateItems))
+			if err != nil {
+				s.lc.Errorf("AGWriteMulti Error: %s, reconnecting...", err)
+				s.mu.Lock()
+				s.s7Clients[deviceName] = nil
+				s.mu.Unlock()
+				s.s7Clients[deviceName] = s.NewS7Client(deviceName, protocols)
+				s7Client = s.s7Clients[deviceName]
+			} else {
+				s.lc.Debugf("AGWriteMulti write from 'dataset': %s", dataset)
+				break
+			}
+
+			retrytimes--
+			if retrytimes == 0 {
+				break
+			}
+		}
+	}
+
+	for _, s7_error := range s7_errors {
+		if s7_error != "" {
+			s.lc.Errorf("S7 Client AGWriteMulti error: %s", s7_error)
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Stop the protocol-specific DS code to shutdown gracefully, or
+// if the force parameter is 'true', immediately. The driver is responsible
+// for closing any in-use channels, including the channel used to send async
+// readings (if supported).
+func (s *Driver) Stop(force bool) error {
+
+	s.mu.Lock()
+	s.s7Clients = nil
+	s.mu.Unlock()
+
+	// Then Logging Client might not be initialized
+	if s.lc != nil {
+		s.lc.Debugf("Driver.Stop called: force=%v", force)
+	}
+	return nil
+}
+
+// AddDevice is a callback function that is invoked
+// when a new Device associated with this Device Service is added
+func (s *Driver) AddDevice(deviceName string, protocols map[string]models.ProtocolProperties, adminState models.AdminState) error {
+	s.lc.Debugf("a new Device is added: %s", deviceName)
+
+	s.s7Clients[deviceName] = nil
+	s7Client := s.NewS7Client(deviceName, protocols)
```

Comment:
Again, you can reduce duplicate fcode for locking and saving the client if you call `getS7Client` instaed.

Reply:
Yes, the code is duplicated here.  You can just call `getS7Client`